rfcs/20200612-stream-executor-c-api.md

```diff
@@ -0,0 +1,557 @@
+  // initialized.
+  TF_BOOL (*device_memory_usage)(
+      SE_Device* executor, int64_t* free, int64_t* total);
+
+
+  /*** STREAM CALLBACKS ***/
+  // Creates SE_Stream. This call should also Allocate stream
+  // resources on the underlying platform and initializes its
+  // internals.
+  void (*create_stream)(SE_Device* executor, SE_Stream*, TF_Status*);
+
+  // Destroys SE_Stream and deallocates any underlying resources.
+  void (*destroy_stream)(SE_Device* executor, SE_Stream stream);
+
+  // Causes dependent to not begin execution until other has finished its
+  // last-enqueued work.
+  TF_BOOL (*create_stream_dependency)(
+      SE_Device* executor, SE_Stream dependent,
+      SE_Stream other);
+
+  // Without blocking the device, retrieve the current stream status.
+  void (*get_status)(SE_Device* executor, SE_Stream stream,
+                     TF_Status* status);
+
+  /*** EVENT CALLBACKS ***/
+  // Create SE_Event. Performs platform-specific allocation and initialization of an event.
+  void (*create_event)(
+      SE_Device* executor, SE_Event* event, TF_Status* status);
+
+  // Destroy SE_Event and perform any platform-specific deallocation and cleanup of an event.
+  void (*destroy_event)(
+      SE_Device* executor, SE_Event event, TF_Status* status);
+
+  // Requests the current status of the event from the underlying platform.
+  SE_EventStatus (*poll_for_event_status)(
+      SE_Device* executor, SE_Event event);
+  // Inserts the specified event at the end of the specified stream.
+  void (*record_event)(
+      SE_Device* executor, SE_Stream stream,
+      SE_Event event, TF_Status* status);
+
+  // Wait for the specified event at the end of the specified stream.
+  void (*wait_for_event)(
+      SE_Device* executor, SE_Stream stream,
+      SE_Event event, TF_Status* status);
+
+  /*** TIMER CALLBACKS ***/
+  // Creates TF_Timer. Allocates timer resources on the underlying platform and initializes its
+  // internals.
+  void (*create_timer)(SE_Device* executor, SE_Timer* timer, SE_TimerFns** timer_fns, TF_Status* status);
+
+  // Destroy timer and deallocates timer resources on the underlying platform.
+  void (*destroy_timer)(SE_Device* executor, SE_Timer timer, SE_TimerFns* timer_fns);
+
+  // Records a start event for an interval timer.
+  TF_BOOL (*start_timer)(
+      SE_Device* executor, SE_Stream stream, SE_Timer timer);
+
+
+  // Records a stop event for an interval timer.
+  TF_BOOL (*stop_timer)(
+      SE_Device* executor, SE_Stream stream, SE_Timer timer);
+
+  /*** MEMCPY CALLBACKS ***/
+  // Entrains a memcpy operation onto stream, with a host destination location
+  // host_dst and a device memory source, with target size size.
+  TF_BOOL (*memcpy_to_host)(
+      SE_Device* executor, SE_Stream stream,
+      void* host_dst,
+      const SE_DeviceMemoryBase* device_src,
+      uint64_t size);
+
+  // Entrains a memcpy operation onto stream, with a device destination location
```

Comment:
I've heard people use "Enqueue" instead.  Any specific reason to prefer Entrain?

Reply:
+1 enqueue would be something that I'm used to as well.

Reply:
The comment was copied from `stream_executor_pimpl.h`. I will update it here, "enqueue" sgtm.

Reply:
Done